File: tests/node/test_fal_compute_smoke.mjs
/**
 * Comprehensive fal.ai compute smoke tests.
 *
 * Gated on the FAL_API_KEY environment variable.
 * Tests image generation, text-to-speech, music generation, transcription,
 * text-to-video, raw compute, and job lifecycle.
 */

import { describe, it } from "node:test";
import assert from "node:assert/strict";

import { FalProvider } from "../../crates/blazen-node/index.js";

const FAL_API_KEY = process.env.FAL_API_KEY;

// -- Image Generation --------------------------------------------------------

describe("fal.ai image generation smoke tests", { skip: !FAL_API_KEY }, () => {
  it("generates an image from a text prompt", async () => {
    const provider = FalProvider.create(FAL_API_KEY);
    const result = await provider.generateImage({
      prompt: "a red circle on a white background",
    });

    assert.ok(result, "expected a result");
    assert.ok(result.images, "expected images in result");
    assert.ok(result.images.length > 0, "expected at least one image");
  });
});

// -- Text to Speech ----------------------------------------------------------

describe("fal.ai text-to-speech smoke tests", { skip: !FAL_API_KEY }, () => {
  it("synthesizes speech from text", async () => {
    const provider = FalProvider.create(FAL_API_KEY);
    const result = await provider.textToSpeech({
      text: "Hello world.",
    });

    assert.ok(result, "expected a result");
    // The result should contain audio data (url or base64)
    assert.ok(
      result.audio_url || result.audio || result.url,
      `expected audio data in result, got keys: ${Object.keys(result).join(", ")}`
    );
  });
});

// -- Music Generation --------------------------------------------------------

describe("fal.ai music generation smoke tests", { skip: !FAL_API_KEY }, () => {
  it("generates music from a prompt", async () => {
    const provider = FalProvider.create(FAL_API_KEY);
    const result = await provider.generateMusic({
      prompt: "happy upbeat jingle",
      durationSeconds: 5,
    });

    assert.ok(result, "expected a result");
    // The result should contain audio data
    assert.ok(
      result.audio_url || result.audio || result.url || result.audio_file,
      `expected audio data in result, got keys: ${Object.keys(result).join(", ")}`
    );
  });
});

// -- Transcription -----------------------------------------------------------

describe("fal.ai transcription smoke tests", { skip: !FAL_API_KEY }, () => {
  it("transcribes audio from a URL", async () => {
    const provider = FalProvider.create(FAL_API_KEY);
    const result = await provider.transcribe({
      audioUrl: "https://cdn.openai.com/API/docs/audio/alloy.wav",
    });

    assert.ok(result, "expected a result");
    // The result should contain transcribed text
    assert.ok(
      result.text || result.transcript || result.chunks,
      `expected text content in result, got keys: ${Object.keys(result).join(", ")}`
    );
  });
});

// -- Text to Video -----------------------------------------------------------

describe("fal.ai text-to-video smoke tests", { skip: !FAL_API_KEY }, () => {
  it("generates a video from a text prompt", { timeout: 300_000 }, async () => {
    const provider = FalProvider.create(FAL_API_KEY);
    const result = await provider.textToVideo({
      prompt: "a cat walking slowly",
    });

    assert.ok(result, "expected a result");
    // The result should contain video data
    assert.ok(
      result.videos || result.video_url || result.video || result.url,
      `expected video data in result, got keys: ${Object.keys(result).join(", ")}`
    );
    if (result.videos) {
      assert.ok(result.videos.length > 0, "expected at least one video");
    }
  });
});

// -- Raw Compute (run) -------------------------------------------------------

describe("fal.ai raw compute smoke tests", { skip: !FAL_API_KEY }, () => {
  it("runs a model synchronously via run()", async () => {
    const provider = FalProvider.create(FAL_API_KEY);
    const result = await provider.run("fal-ai/flux/schnell", {
      prompt: "blue sky with white clouds",
      image_size: "square_hd",
    });

    assert.ok(result, "expected a non-null result from run()");
  });
});

// -- Job Lifecycle (submit + status) -----------------------------------------

describe("fal.ai job lifecycle smoke tests", { skip: !FAL_API_KEY }, () => {
  it("submits a job and gets a valid job handle", async () => {
    const provider = FalProvider.create(FAL_API_KEY);

    const job = await provider.submit("fal-ai/flux/schnell", {
      prompt: "green forest with sunlight",
    });

    assert.ok(job, "expected a job handle from submit()");

    const jobId = job.id || job.requestId || job.request_id;
    assert.ok(jobId, `expected job id, got keys: ${Object.keys(job).join(", ")}`);
    assert.ok(typeof jobId === "string" && jobId.length > 0, "job id should be a non-empty string");
  });
});

File: tests/python/test_fal_compute_smoke.py
"""fal.ai compute smoke tests: image, video, audio, transcription, job lifecycle.

Gated on the FAL_API_KEY environment variable.
"""

import os

import pytest

from blazen import FalProvider

FAL_API_KEY = os.environ.get("FAL_API_KEY")

skip_without_key = pytest.mark.skipif(
    not FAL_API_KEY,
    reason="FAL_API_KEY not set",
)


# ---------------------------------------------------------------------------
# Image generation
# ---------------------------------------------------------------------------


@skip_without_key
@pytest.mark.asyncio
async def test_fal_generate_image():
    """Generate an image and verify the response contains image data."""
    provider = FalProvider(api_key=FAL_API_KEY)
    result = await provider.generate_image(
        {"prompt": "a red circle on white background"}
    )

    assert "images" in result
    assert len(result["images"]) > 0


# ---------------------------------------------------------------------------
# Text-to-speech
# ---------------------------------------------------------------------------


@skip_without_key
@pytest.mark.asyncio
async def test_fal_text_to_speech():
    """Synthesize speech and verify the response contains audio data."""
    provider = FalProvider(api_key=FAL_API_KEY)
    result = await provider.text_to_speech(
        {"text": "Hello world."}
    )

    assert "audio" in result
    assert len(result["audio"]) > 0


# ---------------------------------------------------------------------------
# Music generation
# ---------------------------------------------------------------------------


@skip_without_key
@pytest.mark.asyncio
async def test_fal_generate_music():
    """Generate music and verify the response contains audio data."""
    provider = FalProvider(api_key=FAL_API_KEY)
    result = await provider.generate_music(
        {"prompt": "happy upbeat jingle", "duration_seconds": 5.0}
    )

    assert "audio" in result
    assert len(result["audio"]) > 0


# ---------------------------------------------------------------------------
# Transcription
# ---------------------------------------------------------------------------


@skip_without_key
@pytest.mark.asyncio
async def test_fal_transcribe():
    """Transcribe a short audio clip and verify text is returned."""
    provider = FalProvider(api_key=FAL_API_KEY)
    result = await provider.transcribe(
        {"audio_url": "https://cdn.openai.com/API/docs/audio/alloy.wav"}
    )

    assert "text" in result
    assert len(result["text"]) > 0


# ---------------------------------------------------------------------------
# Text-to-video (slow -- extended timeout)
# ---------------------------------------------------------------------------


@skip_without_key
@pytest.mark.asyncio
@pytest.mark.timeout(120)
async def test_fal_text_to_video():
    """Generate a video from text (slow, ~30-60s)."""
    provider = FalProvider(api_key=FAL_API_KEY)
    try:
        result = await provider.text_to_video(
            {"prompt": "a cat walking"}
        )
    except Exception as e:
        err = str(e).lower()
        if any(
            marker in err
            for marker in (
                "downstream_service_unavailable",
                "service unavailable",
                "service_unavailable",
            )
        ):
            pytest.skip(f"fal.ai video service transiently unavailable: {e}")
        raise

    assert "videos" in result
    assert len(result["videos"]) > 0


# ---------------------------------------------------------------------------
# Raw compute run
# ---------------------------------------------------------------------------


@skip_without_key
@pytest.mark.asyncio
async def test_fal_raw_compute_run():
    """Run a raw compute job via the generic run() method."""
    provider = FalProvider(api_key=FAL_API_KEY)
    result = await provider.run(
        model="fal-ai/flux/schnell",
        input={"prompt": "blue sky", "image_size": "square_hd"},
    )

    assert result is not None
    assert "output" in result or "images" in result


# ---------------------------------------------------------------------------
# Job lifecycle: submit + status
# ---------------------------------------------------------------------------


@skip_without_key
@pytest.mark.asyncio
async def test_fal_job_submit():
    """Submit a job and verify a job handle is returned."""
    provider = FalProvider(api_key=FAL_API_KEY)

    job = await provider.submit(
        model="fal-ai/flux/schnell",
        input={"prompt": "green forest"},
    )

    assert "id" in job
    assert isinstance(job["id"], str)
    assert len(job["id"]) > 0
